### DISCUSSION TITLE
Multiplayer

### DISCUSSION THREAD

**Discussion by randallwc** [ECE-180D-WS-2022/Team3#48] [Ideas]
We are making multiplayer

**Comment by randallwc** [comment]
things to remember

1. we need to have all the enemies be in the same position across all devices
2. we need to have it work for `N` connected players

things that must be shared
1. we need to share the `x,y,z` position of each enemy and each opp ranger
2. we need to share the health of each enemy and opp ranger
3. we need to share the score of each ranger
4. we need to share the powerups across all devices (Powerups not created yet)
5. we need to share the bullet entities across all devices (bullet entity not created yet)

my suggested way to do this
- one person will host
- other people will connect to that room
- the host's game will dictate each enemy step (i.e. movement on the screen each clock tick)
- the host's game will send the location of
  1. all the enemies' locations `x,y,z`
  2. all enemy healths
  3. all the bullet locations `x,y,z` and speed (not implemented yet)
  4. all the powerup locations `x,y,z` (not implemented yet)
- the non-host players will send
  1. when they are firing (T or F)
  2. when they shoot a bullet and its speed (not implemented)
  3. their health
  4. their score
- the non-host players will then receive all the data and render all of it for themselves

**Comment by jcr7467** [comment]
Send what enemy was just killed by python client and not handle it from host, everyone just reacts to it

**Comment by jcr7467** [comment]
from client send to server, 'i killed this dude' and the server will handle that

**Comment by jcr7467** [comment]
always send z level

**Comment by randallwc** [comment]
send clock from host

**Comment by jcr7467** [comment]
i agree

**Comment by jcr7467** [comment]
add username to server

**Comment by randallwc** [comment]
sync up firing

**Comment by randallwc** [comment]
publish num_z_levels on startup i.e. num_z_levels=3
